[FLINK-19082][doc] Add docs for temporal table and temporal table join (#14152)
apache/flink: docs/dev/table/streaming/temporal_tables.md

KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
--> 

> Temporal table is a table that evolves over time as known as Flink [dynamic table](dynamic_tables.html), rows in temporal table are associated with one or more temporal periods, all Flink tables are temporal(dynamic) table.

Reviewer: ```suggestion
A temporal table is a table that evolves over time -  otherwise known in Flink as a [dynamic table](dynamic_tables.html). Rows in a temporal table are associated with one or more temporal periods and all Flink tables are temporal(dynamic).
```

A temporal table contains one or more versioned table snapshots, it can be a changing history table which tracks the changes(e.g. database changelog, contains all snapshots) or a changing dimensioned table which materializes the changes(e.g. database table, contains the latest snapshot). 

**Version**: A temporal table can split into a set of versioned table snapshots, the version in table snapshots represents the valid life circle of rows, the start time and the end time of the valid period can be assigned by users. 
Temporal table can split to `versioned table` and `regular table` according to the table can tracks its history version or not.

> **Versioned table**: If the row in temporal table can track its history changes and visit its history versions, we call this kind of temporal table as versioned table, the table comes from database changelog can define as a versioned table.
> 
> **Regular table**: If the row in temporal table can only track and visit its latest version，we call this kind of temporal table as regular table. The table comes from database or HBase can define as a regular table.

Reviewer: ```suggestion
**Versioned table**: If the rows in a temporal table can track its history changes and visit its history versions, we call this a versioned table. Tables that comes from a database changelog can be defined as a versioned table.

**Regular table**: If the rows in a temporal table can only track and visit its latest version, we call this a regular table. Tables that comes from a database or HBase can be defined as a regular table.
```

Yen (which has a rate of 1). For example, the exchange rate for the period from 09:00 to 10:45 of Euro to Yen was 114.
From 10:45 to 11:15 it was 116.

{% highlight sql %}
SELECT * FROM RatesHistory;

currency_time currency  rate
============= ========= ====
09:00:00      US Dollar 102
09:00:00      Euro      114
09:00:00      Yen       1
10:45:00      Euro      116
11:15:00      Euro      119
11:49:00      Pounds    108
{% endhighlight %}

To define a versioned table on `RatesHistory`, Flink supports defining a versioned view 
by [deduplication query]({{ site.baseurl }}/dev/table/sql/queries.html#deduplication) which produces an ordered changelog
stream with an inferred primary key(`currency`) and event time(`currency_time`).

{% highlight sql %}
-- Define a versioned view
CREATE VIEW versioned_rates AS              
SELECT currency, rate, currency_time            -- (1) `currency_time` keeps the event time
  FROM (
      SELECT *,
      ROW_NUMBER() OVER (PARTITION BY currency  -- (2) the inferred unique key `currency` can be a primary key
         ORDER BY currency_time DESC) AS rowNum 

>       FROM rates )

Reviewer 2: Make sure the examples are "runnable" We should use the `RatesHistory` table here.
